Port to Paper (for 0.101.3.0 || 0.102.0.0)

TheFlagCourier opened the pull request:
This PR lays the groundwork for future paper-native adaptions, given the next version of Towny is planned to only support Paper.

Spigot-API has been swapped for Paper-API 1.19.4 (This can probably be raised, playing it safe.)

Dependencies have been updated, and the JVM language level has been upgraded to Java 17. The minimum supported Towny version has been bumped to 0.101.2.1.

Also in this PR is some basic house-cleaning:
- README adapted from Towny's README
- LICENSE copied from Towny.
- Removed Jitpack and Jabel

TheFlagCourier committed:
Port to Paper (for 0.101.3.0 || 0.102.0.0); Iterate towards next 0.120

- Update dependencies
- Target Java 17
- Target 0.101.2.1 as minimum Towny version
- Port Towny's `README.md`
- Add `LICENSE.md` from Towny

TheFlagCourier commented:
@LlmDl The ported README will need some review, but  otherwise I think this should be alright. Let me know if I need to make any changes.

TheFlagCourier committed:
Add missing provided scope for EssentialsXDiscord

Okay, not sure how I messed this up...

LlmDl reviewed (commented):
Looks quite nice, thank you. This will go ahead once Towny goes full paper.

TheFlagCourier commented:
Sounds good. If I port anything else (e.g. AsyncPlayerChatEvent -> AsyncChatEvent), and this is still incubating, I'll target this branch for PRs.

Warriorrrr commented on README.md:
```diff
@@ -0,0 +1,192 @@
+
+### Building TownyChat
+If you would like to build TownyChat yourself, you can do so with [Apache Maven](https://maven.apache.org/).
+Depending on the age of the branch, you may wish to instead look at [Apache Ant](https://ant.apache.org).
+Building with Ant is outside of this brief tutorial's scope.
```
TownyChat doesn't have any branches that require ant so it's probably not worth mentioning here

TheFlagCourier replied:
Probably not, yeah. Only really applies if they checkout a specific commit.

<img width="1400" height="1000" alt="image" src="https://github.com/user-attachments/assets/8c47465f-9a95-4496-9456-141526927408" />

Warriorrrr commented on pom.xml:
```diff
@@ -83,56 +77,99 @@
+            <groupId>io.papermc.paper</groupId>
+            <artifactId>paper-api</artifactId>
+            <version>1.19.4-R0.1-SNAPSHOT</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>com.google.guava</groupId>
+                    <artifactId>guava</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.yaml</groupId>
+                    <artifactId>snakeyaml</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>net.kyori</groupId>
+                    <artifactId>adventure-text-serializer-gson</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.apache.maven</groupId>
+                    <artifactId>maven-resolver-provider</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>net.md-5</groupId>
+                    <artifactId>bungeecord-chat</artifactId>
+                </exclusion>
+            </exclusions>
```
What's up with all these exclusions?

TheFlagCourier replied:
That's just me getting IDEA to shut up about vulnerable transitives (even though they are not in the compile scope.)

Builds fine without them, so I just left them excluded. The whole exclusions block can be dropped, otherwise.

TheFlagCourier replied:
I've removed all exclusions except for emoji-java (EssXDiscord) as it would require Jitpack to be declared.

Warriorrrr commented on pom.xml:
```diff
@@ -83,56 +77,99 @@
             </exclusions>
         </dependency>
         <dependency>
             <groupId>us.dynmap</groupId>
             <artifactId>dynmap-api</artifactId>
-            <version>3.0-SNAPSHOT</version>
+            <version>${dynmap.version}</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>us.dynmap</groupId>
+            <artifactId>DynmapCoreAPI</artifactId>
+            <version>${dynmap.version}</version>
             <scope>provided</scope>
         </dependency>
```
Having two dynmap dependencies also seems weird

TheFlagCourier replied:
It seems weird, yes. But it's due to the present implementation, and Dynmap's `dynmap-api` package not including `DynmapCoreAPI`, transitively.

- Comment out `dynmap-api`, and the symbol for `org.dynmap.DynmapAPI` is gone.

- Comment out `DynmapCoreAPI` and the symbol for `org.dynmap.DynmapCommonAPI` is missing.

TheFlagCourier committed:
README.md - Remove mention of ANT build tools

As pointed out by @Warriorrr, this is pointless to mention. There are no stale branches or tags that point to ANT-reliant versions.

I'm sure someone manually checking out from a commit signature, is smart enough to figure it out.

TheFlagCourier committed:
POM - Removed most exclusions for transitive libraries

Not actively overriding or depending on any of these libraries. These are still exposed at runtime regardless.

Exclusion for `emoji-java` kept as it is otherwise unused and only available through JitPack.

Warriorrrr commented on src/com/palmergames/bukkit/TownyChat/Chat.java:
```diff
@@ -58,7 +60,7 @@ public class Chat extends JavaPlugin {
 	private Essentials essentials = null;
 	
-	private static String requiredTownyVersion = "0.100.4.0";
+	private static final String requiredTownyVersion = "0.101.2.1";
```
Can be changed to 0.101.2.5 now that it's out, it's the first version that depends on Paper.

TheFlagCourier replied:
Late response, but I've bumped it to 0.102.0.0 instead, since it's the current "stable" release.

TheFlagCourier committed:
Bump reqTownyVersion to 0.102.0.0

TheFlagCourier committed:
Bump versions

LlmDl commented on README.md:
```diff
@@ -0,0 +1,190 @@
+> Prior versions of TownyChat may continue to function, but are unsupported if not used as advertised.
+
+* **[Current Development](https://github.com/TownyAdvanced/TownyChat) \[Future 0.120\]**
+  * Supports Towny 0.101.2.1+ (Pre-Release)
```
```suggestion
  * Supports Towny 0.102.0.0+
```

LlmDl commented on pom.xml:
```diff
@@ -23,10 +23,12 @@
     <properties>
-        <java.version>17</java.version>
-        <project.bukkitAPIVersion>1.16</project.bukkitAPIVersion>
+        <java.version>21</java.version>
```
```suggestion
        <java.version>17</java.version>
```

LlmDl replied:
1.19 is still supported and can be run on java 17.

LlmDl commented on pom.xml:
```diff
@@ -23,10 +23,12 @@
-        <java.version>17</java.version>
-        <project.bukkitAPIVersion>1.16</project.bukkitAPIVersion>
+        <java.version>21</java.version>
+        <project.bukkitAPIVersion>1.21</project.bukkitAPIVersion>
```
```suggestion
        <project.bukkitAPIVersion>1.19</project.bukkitAPIVersion>
```

LlmDl replied:
This value gets dropped into the plugin.yml at compile, and would prevent anyone on 1.19-1.20 from running the plugin.

LlmDl committed:
Switch up some numbers
